public/extensions/breakout.js

```diff
@@ -0,0 +1,116 @@
+// @ts-check
+
+import daily from "./core.js";
+import Socket from "./socketiostate.js";
+
+let call;
+let socket;
+
+// Initial state
+let state = {
+  initialRoomUrl: "",
+  breakoutStarted: false,
+  participants: [{ user_name: "", roomUrl: "" }],
+};
+
+daily.beforeCreateFrame((parentEl, properties) => {
+  if (!properties.customTrayButtons) {
+    properties.customTrayButtons = {};
+  }
+
+  properties.customTrayButtons = {
+    ...properties.customTrayButtons,
+    toggleBreakout: {
+      iconPath: "https://www.svgrepo.com/show/207394/flash.svg",
+      label: "Breakout",
+      tooltip: "Breakout",
+    },
+  };
+
+  return [parentEl, properties];
+});
+
+daily.afterCreateFrame(async (c) => {
+  call = c;
+  call.on("custom-button-click", (e) => {
+    if (e.button_id !== "toggleBreakout") {
+      return;
+    }
+
+    state.breakoutStarted ? end() : start();
+  });
+});
+
+async function handleOnStateUpdate(s = {}) {
+  state = { ...state, ...s };
+
+  const localParticipant = call.participants().local;
+
+  const roomUrl = state.participants.find(
+    (p) => p.user_name === localParticipant?.user_name
+  )?.roomUrl;
+
+  if (roomUrl) {
+    try {
+      await call.leave();
+      await call.join({ url: roomUrl });
+    } catch (err) {
+      console.error("Failed to join new room:", err);
+    }
+  }
+}
+
+export function connect({ room = "", domain = "" }) {
+  const key = `${domain}/${room}/breakout`;
+  state.initialRoomUrl = `https://${domain}.daily.co/${room}`;
+  socket = new Socket({ key });
+  socket.onStateUpdate(handleOnStateUpdate);
+  socket.connect();
+}
+
+function randomizeParticipants(participants = {}, roomUrls = []) {
+  return Object.entries(participants)
+    .map((value) => ({ value, sort: Math.random() })) // shuffle participants
+    .sort((a, b) => a.sort - b.sort)
+    .map(({ value }) => value)
+    .map(([_, participant], index) => {
+      // Evenly split into two rooms
+      const roomUrl = index % 2 === 0 ? roomUrls[0] : roomUrls[1];
+      return { roomUrl, user_name: participant.user_name };
+    });
+}
+
+export async function start() {
+  const response = await fetch("/create-rooms", {
+    method: "POST",
+    headers: {
+      Accept: "application/json",
+      "Content-Type": "application/json",
+    },
+  });
+
+  const { roomUrls } = await response.json();
+
+  // Randomize participants
+  const participants = randomizeParticipants(call.participants(), roomUrls);
+
+  // Send state to all clients
+  socket.updateState({
+    participants,
+    breakoutStarted: true,
+  });
+}
+
+export function end() {
+  // 1. Get a list of participants from the server state
+  // and bring them all back to the main room
+
+  const participants = state.participants.map((participant) => {
+    return { roomUrl: state.initialRoomUrl, user_name: participant.user_name };
+  });
+
+  socket.updateState({
+    participants,
+    breakoutStarted: false,
```

Comment:
Thinking out loud here, and this might be a bit too clever... but we could eliminate the need for a separate `breakoutStarted` boolean by checking for the presence/absence of entries in the `participants` array itself. I don't feel too strongly about it now, but maybe something to consider for the future—especially as it relates to some of my ideas around making this a Proper Extension that works for Hey Lady and others. :)

In fact, you could store the internal state like this when a breakout is happening:

```
{
  participants: [
    {"guid1": "https://domain.daily.co/room-1"},
    {"guid2": "https://domain.daily.co/room-2"}
  ],
  default: "https://domain.daily.co/main-room"
}
```

And then this when it isn't:
```
{
  participants: [],
  default: "https://domain.daily.co/main-room"
}
```

Then, at any moment, a participant can put themself into the right room by doing the `find` on line 49, and if they don't find themselves, they go to the default room.